akka-stream/src/main/scala/akka/stream/KillSwitch.scala

```diff
@@ -0,0 +1,259 @@
+/**
+ * Copyright (C) 2015-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream
+
+import akka.Done
+import akka.stream.stage._
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * Creates shared or single kill switches which can be used to control completion of graphs from the outside.
+ *  - The factory ``shared()`` returns a [[SharedKillSwitch]] which provides a [[Graph]] of [[FlowShape]] that can be
+ *    used in arbitrary number of graphs and materializations. The switch simultaneously
+ *    controls completion in all of those graphs.
+ *  - The factory ``single()`` returns a [[Graph]] of [[FlowShape]] that materializes to a [[UniqueKillSwitch]]
+ *    which is always unique
+ *    to that materialized Flow itself.
+ *
+ * Creates a [[SharedKillSwitch]] that can be used to externally control the completion of various streams.
+ *
+ */
+object KillSwitches {
+
+  /**
+   * Creates a new [[SharedKillSwitch]] with the given name that can be used to control the completion of multiple
+   * streams from the outside simultaneously.
+   *
+   * @see SharedKillSwitch
+   */
+  def shared(name: String): SharedKillSwitch = new SharedKillSwitch(name)
+
+  /**
+   * Creates a new [[Graph]] of [[FlowShape]] that materializes to an external switch that allows external completion
+   * of that unique materialization. Different materializations result in different, independent switches.
+   *
+   * For a Bidi version see [[KillSwitch#singleBidi]]
+   */
+  def single[T]: Graph[FlowShape[T, T], UniqueKillSwitch] =
+    UniqueKillSwitchStage.asInstanceOf[Graph[FlowShape[T, T], UniqueKillSwitch]]
+
+  /**
+   * Creates a new [[Graph]] of [[FlowShape]] that materializes to an external switch that allows external completion
+   * of that unique materialization. Different materializations result in different, independent switches.
+   *
+   * For a Flow version see [[KillSwitch#single]]
+   */
+  def singleBidi[T1, T2]: Graph[BidiShape[T1, T1, T2, T2], UniqueKillSwitch] =
+    UniqueBidiKillSwitchStage.asInstanceOf[Graph[BidiShape[T1, T1, T2, T2], UniqueKillSwitch]]
+
+  abstract class KillableGraphStageLogic(val terminationSignal: Future[Done], _shape: Shape) extends GraphStageLogic(_shape) {
+    override def preStart(): Unit = {
+      terminationSignal.value match {
+        case Some(status) ⇒ onSwitch(status)
+        case _ ⇒
+          // callback.invoke is a simple actor send, so it is fine to run on the invoking thread
+          terminationSignal.onComplete(getAsyncCallback[Try[Done]](onSwitch).invoke)(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+      }
+    }
+
+    private def onSwitch(mode: Try[Done]): Unit = mode match {
+      case Success(_)  ⇒ completeStage()
+      case Failure(ex) ⇒ failStage(ex)
+    }
+  }
+
+  private[stream] object UniqueKillSwitchStage extends GraphStageWithMaterializedValue[FlowShape[Any, Any], UniqueKillSwitch] {
```

Comment:
> didn't we discover that nested objects involved some volatile (or something)?

That case was nested objects in a class which were used all the time (old interpreter). Probably a val is cheaper here but I don't think this is a hot path, especially since the stage is reusable.